### DISCUSSION TITLE
Redirect & Browser Back Form Interaction Bugs

### DISCUSSION THREAD

**Discussion by M-Porter** [livewire/livewire#3063] [7. Bugs (archived)]
### Description

There are actually 2 bugs here related to form behaviors after navigating back. Not opening two separate tickets as I believe they are related.

* **Bug 1** – When navigating back in the browser (clicking the back button) after doing an action which causes a redirect, forms are no longer editable. When inspecting the DOM, a `readonly=""` attribute remains on the input element. 

* **Bug 2** –When navigating back in the borwser (clicking the back button) after doing an action which causes a redirect, the form input is editable but clicking submit does nothing. The form goes into a read-only and disabled state but no xhr request gets sent.

Only a full page refresh (as expected) fixes this.

Webapp of bug can be found at https://livewire-redirect-back-ro-bug.herokuapp.com/
Code repository of bug can be found at https://github.com/M-Porter/livewire-redirect-back-readonly-bug

### Exact steps to reproduce

To make it easier i setup a heroku instance to run this bug:

**Bug 1**
1. Go to https://livewire-redirect-back-ro-bug.herokuapp.com/
2. Enter a value in the input field and click submit.
3. After being redirected to the new page, click the back button in your browser.
4. See your original input value in the input field. The input field is not editable.

**Bug 2**
1. Go to https://livewire-redirect-back-ro-bug.herokuapp.com/
2. Enter a value in the input field and click submit.
3. After being redirected to the new page, click the back button in your browser.
4. See your original input which is editable.
5. Click submit, nothing happens (no xhr).

### Stripped-down, copy-pastable code snippets

I'm not really sure which code snippets to include (which is why i made the repo to follow, but here are the files i changed from the base laravel install:
* Routes – https://github.com/M-Porter/livewire-redirect-back-readonly-bug/blob/main/routes/web.php
* Controller – https://github.com/M-Porter/livewire-redirect-back-readonly-bug/blob/main/app/Http/Controllers/Controller.php
* Livewire Component – https://github.com/M-Porter/livewire-redirect-back-readonly-bug/blob/main/app/Http/Livewire/SomeForm.php
* Views – https://github.com/M-Porter/livewire-redirect-back-readonly-bug/tree/main/resources/views

**Context**
* Livewire version: v2.4.0
* Laravel version: v8.31.0
* Alpine version: N/A
* Browser: Chrome, Firefox, Safari

**Comment by M-Porter** [comment]
If you change the `some-form` component to no longer use a form but instead attach `wire:keydown.enter="submit"` to the input and `wire:click="submit"` to the button, the code in [`DisableForms.js`](https://github.com/livewire/livewire/blob/73a790be20890eddc99ca697b80415ce1a1b3ca8/js/component/DisableForms.js) is no longer fired which actually solves Bug 1. Meaning bug 1 is due to bad cleanup or lack of proper hooks when navigating.

Bug 2 however still occurs. After navigating back no form updates are made: no DOM diffs, no input validation. No `wire:` action xhr requests occur.

**Comment by joshhanley** [comment]
@M-Porter yeah there are issues with back navigation and browser cache, check out #2114 and #2334 and see if any of the solutions there can help.

**Comment by M-Porter** [comment]
The full page reload when looking for `event.persisted` worked as expected. I stated as much in my bug report.

When using a basic form with no livewire, this issue does not occur. As far as I am concerned this is a regression in a standard behavior and should be addressed by the livewire team.

The only way i figured out how to get around the issue was by having the following as my livewire form component:
```html
<form
    action="/some-endpoint"
    method="GET"
>
    <div>
        <input
            type="text"
            wire:model="someModel"
        >
        @error('someModel')
            <div class="errors">
                {{ $message }}
            </div>
        @enderror
    </div>
    <button
        type="submit"
    >
        Search
    </button>
</form>
```

And then in the handler for `/some-endpoint`, use a custom form request with the same rules as the Livewire component rules.

So pretty much, livewire can only be used for live form validation but not for the form submission.

**Edit:** 

Given that normal back behavior works correctly and the offenders, the `readonly=` and `disabled=` attributes, are added via javascript this is not a browser cache issue but Livewire bug. Livewire's hooks are unable to pickup where it left off effectively. This is not an issue in either Rails, Phoenix, or Stimulus so if they can solve for it, so can Livewire.

**Comment by M-Porter** [comment]
Closing as it seems the livewire team has no interest in solving this issue.

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by joshhanley** [comment]
@M-Porter I'm sorry you feel that way. But I'd like to know, what gave you that impression that there is no interest in fixing?

Please understand that there is no Livewire team, there is just Caleb and a bunch of volunteers, like myself, who put their spare time into helping out where they can. But most of us have full time jobs and other stuff to contend with, so available time can be limited.

This problem is such a deep problem, it's not something that can be fixed over night. It goes deep into how the browsers act when it comes to back button handing and browser cache, to the point that may need a deeper, more fundamental architecture change in Livewire to truly solve.

But as there is no clear answer, it is hard to even work through the problem. Even more so when volunteers have limited time.

My time at the moment is mostly limited to weekends, for looking into deep problems like this, so I generally don't comment on these bigger issue until I've had time to go through it in detail. Or if I can find a quick win for you by linking to other threads (like in my original comment).

I hope you understand.

Regards,
Josh

**Comment by M-Porter** [comment]
I do understand. Please don't misinterpret what i said for calling you or the volunteers lazy or not caring about the project. Its clear you all do. The reason i said this is its a major regression from basic browser behavior which other frameworks have already found solutions for _before_ their v1 releases. Livewire is a v2 library being heavily pushed by Laravel as being a feature complete drop-in for creating remarkable applications. In your linked issues, there was no conversation about the possibility of fixing a regression that only seems to affect livewire but rather "not our problem" (obviously paraphrasing of how i feel, not literally what you said). If there is no clear solution, then the website docs should be updated to be honest will ALL readers wishing to use livewire that using forms will not work as expected outside a small subset of use cases and may even cause unintended user interactions.

**Comment by bastinald** [comment]
This is a huge deal.

When someone modifies state on a page and the clicks a link, and hits their browser back button, the state reflected is the original state of the component before any changes were made.

This is confusing as hell and the only solution seems to be a full page refresh after certain state is modified. Which defeats the entire point behind livewire to begin with.

**Comment by calebporzio** [comment]
I'm reopening this, so we can all look deeper into this issue.

Thanks for handling that so kindly @joshhanley.

This is an open source project. Everyone is donating their time out of good will. Being accusatory and feeling entitled is a major turn off for people helping out. Kindness goes a long way.

Let's flip the tone of this issue and discuss what the root problems are, how to reproduce them, and potential solutions.

@M-Porter and @bastinald , have you had a chance to fork the repo and explore the problem?

**Comment by bastinald** [comment]
The issue is that mutated state isn't observed via live wire in order to update the actual browser state when the dom is morphed. 

That's why hitting the back button reverts to the initial component state before changes were made. 

Google MutationObserver for more info.

**Comment by M-Porter** [comment]
I was able to get two hacky solutions working... the first was  to make sure the disabled and readonly states were removed prior to redirect OR when the page loads back up loop over and remove all readonly or disabled attributes if no loading events were taking place. The second option seems extremely costly because its basically akin to cleaning patching hole in a boat after its already sunk.
